lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,203 @@
+    end
+
+    def self.uri_scheme
+      nil
+    end
+
+    # :call-seq:
+    #   add( remote_uri )                       { |input_writer| ... }
+    #   add( remote_uri, byte_count )           { |input_writer| ... }
+    #
+    #   add( local_io, remote_uri )
+    #   add( local_io, remote_uri, byte_count )
+    #
+    # Add a file to the destination URI.
+    #
+    # In the block form of the method, only the remote_uri is required, and it
+    # is assumed the input will be a generated in the executed block (most
+    # likely an external process) to a unix pipe that can be written to.  The
+    # pipe generated by this method and passed in to the block as a file
+    # location to the `input_stream`).
+    #
+    # In the non-block form, a source must be provioded as the first argument
+    # either as an IO object that can be read from, or a file path, and the
+    # second argument is the remote_uri as in the block form.
+    #
+    # An additional argument in both forms as the last argument is `byte_count`
+    # can also be included.  If passed, it will be assumed that the resulting
+    # input will be split, and the naming for the splits will be:
+    #
+    #   - filename.00001
+    #   - filename.00002
+    #   ...
+    #
+    # Block form:
+    #
+    #   nfs_session.add("path/to/file", "200M") do |input_stream|
+    #     `pg_dump -f #{input_stream} vmdb_production`
+    #   end
+    #
+    # Non-block form:
+    #
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file")
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file", "200M")
+    #
+    def add(*upload_args, &block)
+      initialize_upload_vars(*upload_args)
+      mkdir(File.dirname(@remote_file_path))
+      handle_io_block(&block)
+      if byte_count
+        upload_splits
+      else
+        upload_single(@remote_file_path)
+      end
+    ensure
+      reset_vars
+    end
+    alias upload add
+
+    def mkdir(dir) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}##{__callee__} is not defined"
+    end
+
+    # :call-seq:
+    #   download( local_io, remote_uri )
+    #   download( nil,      remote_uri ) { |input_writer| ... }
+    #
+    # Download a file from a remote uri.
+    #
+    # In non-block form, the remote_uri is saved to the local_io.
+    #
+    # In block form, the local_io is omitted, and it is set to a PTY writer
+    # path that will assumed to be read by the block provided.
+    def download(local_file, remote_file_uri, &block)
+      @remote_file_path = remote_file_uri
+      if block_given?
+        thread = handle_io_block(&block)
+        download_single(remote_file_uri, input_writer)
+        input_writer.close
+        thread.join
+      else
+        download_single(remote_file_uri, local_file)
+      end
+    ensure
+      reset_vars
+    end
+
+    private
+
+    def upload_single(file) # rubocop:disable Lint/UnusedMethodArgument
```

Comment:
Can you put a comment here describing how this should be implemented at a high level to work with splitting?

Reply:
Yeah, good call.

I will probably point to one of the existing classes for examples, but yeah, this should definitely be documented.